File: src/object/wallet_inspector.cpp
#include "object/wallet_inspector.h"
#include "locale/locale_manager.h"

bool WalletInspector::seeded = false;

// 统一的可用纸币面额，按金额从大到小排列。
const std::array<int, 6> WalletInspector::denominations = {100, 50, 20, 10, 5, 1};
// 每个情境下提供多条台词，实际使用时随机挑一条。
const std::array<std::string, 3> WalletInspector::zeroWalletMessages = {
    "钱包空空的，一打开只剩下风吹过。",
    "口袋拍了拍，只剩点空气作伴。",
    "翻遍钱包，结果只有比心还干净的底。"};
const std::array<std::string, 3> WalletInspector::summaryLeads = {
    "随手一翻，看到这些钞票：",
    "掀开钱包盖，里面排着：",
    "哗啦啦翻看钱包，里面静静躺着："};
const std::array<std::string, 2> WalletInspector::hasMoreMessages = {
    "剩下的零票得仔细点点才放心。",
    "还有一些散票，想看的话不妨数一遍。"};
const std::array<std::string, 3> WalletInspector::optionCloseMessages = {
    "好吧，先把这些钞票塞回去。",
    "收好钱包，别让风再偷走什么。",
    "合上钱包，留着以后慢慢花。"};
const std::array<std::string, 2> WalletInspector::detailLeadMessages = {
    "仔细点点，所有面额如下：",
    "认真清点一遍，钱包里其实是这样："};
const std::array<std::string, 2> WalletInspector::detailEndMessages = {
    "点好啦，按回车把它们收回去。",
    "数得清清楚楚，回车收好钱包。"};
// 面额对应的形容词，帮助输出更生动的文案。
const std::array<std::string, 6> WalletInspector::denominationDescriptions = {
    "鲜红的",
    "紫色的",
    "翠绿的",
    "蓝灰的",
    "褐色的",
    "浅绿色的"};

WalletInspector::Summary WalletInspector::BuildSummary(const std::array<int, 6> &counts) const
{
    Summary summary;
    summary.hasMore = false;
    // 统一口径：按非零面额种类判断是否空。
    if (CountNonZero(counts) == 0)
    {
        return summary;
    }

    // 直接根据当前钱包面额列出摘要。
    constexpr int kSummaryLimit = 5;
    for (std::size_t i = 0; i < denominations.size(); ++i)
    {
        const int count = counts[i];
        if (count == 0)
        {
            continue;
        }
        summary.lines.emplace_back(FormatEntryAt(i, count));
        if (static_cast<int>(summary.lines.size()) == kSummaryLimit)
        {
            break;
        }
    }
    summary.hasMore = CountNonZero(counts) > static_cast<int>(summary.lines.size());
    return summary;
}

WalletInspector::Detail WalletInspector::BuildDetail(const std::array<int, 6> &counts) const
{
    Detail detail;
    bool any = false;
    for (std::size_t i = 0; i < denominations.size(); ++i)
    {
        const int count = counts[i];
        if (count == 0)
        {
            continue;
        }
        any = true;
        detail.lines.emplace_back(FormatEntryAt(i, count));
    }
    if (!any)
    {
        LocaleManager *locale = LocaleManager::GetInstance();
        detail.lines.emplace_back(locale->Get("wallet.misc", "no_bills"));
    }
    return detail;
}

std::string WalletInspector::PickZeroWalletMessage() const
{
    LocaleManager *locale = LocaleManager::GetInstance();
    auto values = locale->GetSectionValues("wallet.empty");
    if (values.empty())
    {
        return PickRandom(zeroWalletMessages);
    }
    EnsureSeed();
    return values[static_cast<std::size_t>(std::rand() % values.size())];
}

std::string WalletInspector::PickSummaryLead() const
{
    LocaleManager *locale = LocaleManager::GetInstance();
    auto values = locale->GetSectionValues("wallet.summary_lead");
    if (values.empty())
    {
        return PickRandom(summaryLeads);
    }
    EnsureSeed();
    return values[static_cast<std::size_t>(std::rand() % values.size())];
}

std::string WalletInspector::PickHasMoreMessage() const
{
    LocaleManager *locale = LocaleManager::GetInstance();
    auto values = locale->GetSectionValues("wallet.has_more");
    if (values.empty())
    {
        return PickRandom(hasMoreMessages);
    }
    EnsureSeed();
    return values[static_cast<std::size_t>(std::rand() % values.size())];
}

std::string WalletInspector::PickCloseMessage() const
{
    LocaleManager *locale = LocaleManager::GetInstance();
    auto values = locale->GetSectionValues("wallet.close");
    if (values.empty())
    {
        return PickRandom(optionCloseMessages);
    }
    EnsureSeed();
    return values[static_cast<std::size_t>(std::rand() % values.size())];
}

std::string WalletInspector::PickDetailLead() const
{
    LocaleManager *locale = LocaleManager::GetInstance();
    auto values = locale->GetSectionValues("wallet.detail_lead");
    if (values.empty())
    {
        return PickRandom(detailLeadMessages);
    }
    EnsureSeed();
    return values[static_cast<std::size_t>(std::rand() % values.size())];
}

std::string WalletInspector::PickDetailEnd() const
{
    LocaleManager *locale = LocaleManager::GetInstance();
    auto values = locale->GetSectionValues("wallet.detail_end");
    if (values.empty())
    {
        return PickRandom(detailEndMessages);
    }
    EnsureSeed();
    return values[static_cast<std::size_t>(std::rand() % values.size())];
}

int WalletInspector::CountNonZero(const Breakdown &counts)
{
    // 统计仍然存在的面额种类数量。
    int total = 0;
    for (int count : counts)
    {
        if (count > 0)
        {
            ++total;
        }
    }
    return total;
}

std::string WalletInspector::FormatEntryAt(std::size_t index, int count)
{
    LocaleManager *locale = LocaleManager::GetInstance();
    const int denomination = denominations[index];

    // 从 LocaleManager 获取面额描述
    std::string desc;
    switch (denomination)
    {
    case 100:
        desc = locale->Get("wallet.denomination", "d100");
        break;
    case 50:
        desc = locale->Get("wallet.denomination", "d50");
        break;
    case 20:
        desc = locale->Get("wallet.denomination", "d20");
        break;
    case 10:
        desc = locale->Get("wallet.denomination", "d10");
        break;
    case 5:
        desc = locale->Get("wallet.denomination", "d5");
        break;
    case 1:
        desc = locale->Get("wallet.denomination", "d1");
        break;
    default:
        desc = denominationDescriptions[index];
    }

    return std::to_string(count) + " 张" + desc + std::to_string(denomination) + " 元";
}

void WalletInspector::EnsureSeed()
{
    if (!seeded)
    {
        std::srand(static_cast<unsigned int>(std::time(nullptr)));
        seeded = true;
    }
}

File: internal/payment/wallet_inspector.go
package payment

import (
	"citylife/internal/locale"
	"fmt"
)

// Summary 钱包摘要
type Summary struct {
	Lines   []string
	HasMore bool
}

// Detail 钱包详情
type Detail struct {
	Lines []string
}

// WalletInspector 钱包检查器
type WalletInspector struct {
	locale *locale.Locale
}

// NewWalletInspector 创建钱包检查器
func NewWalletInspector(loc *locale.Locale) *WalletInspector {
	return &WalletInspector{
		locale: loc,
	}
}

// BuildSummary 生成摘要（最多5项）
func (w *WalletInspector) BuildSummary(counts [6]int) Summary {
	summary := Summary{
		Lines:   make([]string, 0),
		HasMore: false,
	}

	if countNonZero(counts) == 0 {
		return summary
	}

	const summaryLimit = 5
	for i := range Denominations {
		count := counts[i]
		if count == 0 {
			continue
		}
		summary.Lines = append(summary.Lines, w.formatEntry(i, count))
		if len(summary.Lines) == summaryLimit {
			break
		}
	}

	summary.HasMore = countNonZero(counts) > len(summary.Lines)
	return summary
}

// BuildDetail 生成完整明细
func (w *WalletInspector) BuildDetail(counts [6]int) Detail {
	detail := Detail{
		Lines: make([]string, 0),
	}

	any := false
	for i := range Denominations {
		count := counts[i]
		if count == 0 {
			continue
		}
		any = true
		detail.Lines = append(detail.Lines, w.formatEntry(i, count))
	}

	if !any {
		detail.Lines = append(detail.Lines, w.locale.Get("wallet.misc", "no_bills"))
	}

	return detail
}

// PickZeroWalletMessage 返回"钱包空空"场景的提示语
func (w *WalletInspector) PickZeroWalletMessage() string {
	return w.locale.GetRandom("wallet.empty")
}

// PickSummaryLead 返回摘要部分的开场句
func (w *WalletInspector) PickSummaryLead() string {
	return w.locale.GetRandom("wallet.summary_lead")
}

// PickHasMoreMessage 返回"还有更多"的句子
func (w *WalletInspector) PickHasMoreMessage() string {
	return w.locale.GetRandom("wallet.has_more")
}

// PickCloseMessage 返回"收好钱包"的收尾句
func (w *WalletInspector) PickCloseMessage() string {
	return w.locale.GetRandom("wallet.close")
}

// PickDetailLead 返回"仔细点点"的开场句
func (w *WalletInspector) PickDetailLead() string {
	return w.locale.GetRandom("wallet.detail_lead")
}

// PickDetailEnd 返回"仔细点点"的收尾句
func (w *WalletInspector) PickDetailEnd() string {
	return w.locale.GetRandom("wallet.detail_end")
}

// formatEntry 格式化单项描述
func (w *WalletInspector) formatEntry(index, count int) string {
	denom := Denominations[index]
	desc := w.getDenominationDesc(denom)
	return fmt.Sprintf("%d 张%s%d 元", count, desc, denom)
}

// getDenominationDesc 获取面额描述
func (w *WalletInspector) getDenominationDesc(denom int) string {
	key := fmt.Sprintf("d%d", denom)
	desc := w.locale.Get("wallet.denomination", key)
	// 如果是fallback值，使用默认颜色
	if desc[0] == '[' {
		switch denom {
		case 100:
			return "鲜红的"
		case 50:
			return "紫色的"
		case 20:
			return "翠绿的"
		case 10:
			return "蓝灰的"
		case 5:
			return "褐色的"
		case 1:
			return "浅绿色的"
		default:
			return ""
		}
	}
	return desc
}

// countNonZero 统计非零面额的数量
func countNonZero(counts [6]int) int {
	total := 0
	for _, count := range counts {
		if count > 0 {
			total++
		}
	}
	return total
}

// FormatPaymentResult 格式化支付结果
func FormatPaymentResult(result Result, loc *locale.Locale) string {
	if !result.Success {
		return loc.Get("purchase.error", "insufficient")
	}

	msg := loc.Get("purchase.action", "success") + "\n"
	msg += loc.Get("purchase.action", "paid") + fmt.Sprintf("¥%d\n", result.Paid)

	changeTotal := result.GetChangeTotal()
	if changeTotal > 0 {
		msg += loc.Get("purchase.action", "change") + fmt.Sprintf("¥%d\n", changeTotal)
	}

	return msg
}

// FormatUsedDenominations 格式化扣款明细
func FormatUsedDenominations(used [6]int) string {
	msg := ""
	for i, count := range used {
		if count > 0 {
			msg += fmt.Sprintf("  %d×%d元", count, Denominations[i])
		}
	}
	return msg
}

// FormatChangeDenominations 格式化找零明细
func FormatChangeDenominations(change [6]int) string {
	msg := ""
	for i, count := range change {
		if count > 0 {
			msg += fmt.Sprintf("  %d×%d元", count, Denominations[i])
		}
	}
	return msg
}
